README.md

```diff
@@ -17,1529 +17,197 @@
 - [UCAN-IPLD]
 - [Varsig]
 
-# 0 Abstract
+# Abstract
 
-UCAN Invocation defines a format for expressing the intention to execute delegated UCAN capabilities, the attested receipts from an execution, and how to extend computation via promise pipelining.
+UCAN Invocation defines a format for expressing the intention to execute delegated UCAN capabilities. UCAN Invocation `Receipt` is a signed assertion of the [Executor] state describing [Result] and [Effect]s of the invocation.
 
 ## Language
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
 
-# 1 Introduction
+# Introduction
 
-> Just because you can doesn't mean that you should
->
-> — Anonymous
 
-UCAN is a chained-capability format. A UCAN contains all of the information that one would need to perform some task, and the provable authority to do so. This begs the question: can UCAN be used directly as an RPC language?
+A `Receipt` is a Invocation of the `/ucan/assert` capability. It represents signed assertion from the [Executor] state describing [Result] and [Effect]s of some task invocation. Receipt is a signed commitment by [Executor] to a state, described by it, within the timeframe of the `Receipt`.
```

Comment:
I've been thinking about this a bit recently. It's possible that `/assert` is too broad, since you can assert _all kinds of things_ including content claims, computation receipts, and (oracle) facts about the world. In this current setup, you need to duck type the `args` field.

How do you feel about making `receipt` a subtype of `assert`? It would have all of the same fields, but you'd also know more about the inner structure. Following the new `cmd` format, this would look like `/ucan/assert/receipt` or `/ucan/assert/ran` or similar.

Reply:
> I've been thinking about this a bit recently. It's possible that `/assert` is too broad, since you can assert _all kinds of things_ including content claims, computation receipts, and (oracle) facts about the world. In this current setup, you need to duck type the `args` field.

I have not called that out explicitly but that was kind of the intention behind making `facts` into a map. The way I was thinking about it is that receipt is just an assertion that specifies two expected `{out, run}` facts, but it should be possible to assert arbitrary that about entity in the `about` field they just aren't required by receipt specification.


> How do you feel about making `receipt` a subtype of `assert`? It would have all of the same fields, but you'd also know more about the inner structure. Following the new `cmd` format, this would look like `/ucan/assert/receipt` or `/ucan/assert/ran` or similar.

I suppose ☝️ in comparison might be useful if you want to say allow asserting `{out, run}` but nothing else 🤔 On the other hand if someone adds additional field even if `cmd` was `/ucan/assert/receipt` it probably still should be considered as a valid receipt.

Either way I don't mind changing `cmd` to `/ucan/assert/receipt` although I do not know if I buy into the argument.

Reply:
> in comparison might be useful if you want to say allow asserting {out, run} but nothing else

I think that we're thinking about the design space here very differently. I see the `cmd` field as filling two roles:
1. Define the semantics of the invocation
2. Establish a common language / "which fields are available to that function call"

Receipts have a particular mechanical function with promises. They're more than an assertion of fact — while you can look at it through that lens, these are very different in functionality.

This is the same distinction as shows up in the age old arguments about structural vs nominal typing. Just because the same structural information exists doesn't necessarily mean that it has the same meaning. The analogy breaks down a but here since with a type system you can move that extra bit of information to compile time, which we can't do due to the distributed nature of the problem.

What I liked about the idea of subtyping is that we could nominally tag with the `cmd` field that "the payload is just some assertion, not something to _run_", and leave the fields blank. You can then subtype for more specific uses.

Reply:
I don't think we're thinking that differently it's just you're hoisting all the types all they way to the invocation and use `cmd` as a discriminant, which I generally agree with. In this instance however I feel like not hoisting the type is a better idea and treating it as somewhat constrained generic makes more sense to me. Here is illustration in typescript syntax


What I'm proposing is something like

```ts
type Assert<Facts extends {[key:string]: unknown}> = {
  cmd: "ucan/assert"
  with: DID
  args: { about: Link, facts: Facts }
}

type Receipt = Assert<{ out: Result, run: Link[] }>

// I could also have something like
type LocationCommitment = Assert<{ site: URL }>
```

Where's what you're suggesting is not to make Assert generic but rather hoist to the `cmd` discriminant 

```ts
type Receipt = {
   cmd: "/ucan/assert/receipt"
   with: DID
   args: { about: Link, out: Result, run Link[] } 
}

type LocationCommitment = {
  cmd: "/ucan/assert/commitment"
  with: DID
  args: { about: Link, site: Link[] }
}
```

I do see value generic via generic route as opposed to hoisted, because I can have a neutral component in the system that does not need to worry about the `Facts` parameter e.g. index things regardless of the concrete type. Doing later with hoisted variant is also possible of course but now I have to maintain list of `cmd`s it should work with or adopt some name-spacing like `ucan/assert/`, but even than I loose on the fact that I want all the variants to have `about` field.

On the flip side I do recognize that generic variant I written up no longer features discriminant which certainly can be a drawback. Which is why I'd prefer something like this (where key is a discriminant)

```ts
type Assert<Facts extends {[key:string]: unknown}> = {
  cmd: "ucan/assert"
  with: DID
  args: { about: Link, facts: Facts }
}

type Receipt = Assert<{ receipt: { out: Result, run: Link[] } }>
type Commitment = Assert<{ commitment: { site: URL } }>
```

And it could be sugared more to your liking as follows (although level of advanced typing makes me uncomfortable)

```ts
type Assert<Key extends string, Facts extends {[key: string]: unknown}> = {
   cmd: `/ucan/assert/${Key}`
   with: DID
   args: { about: Link } & Facts
}

type Receipt = Assert<"receipt", { out: Result, run: Link[] }>
type Commitment = Assert<"commitment", { site: URL }>
```

I find last two pretty much identical except later seems to reduce nesting and require more powerful type system and one before is less demanding on type system at the expense of bit more nesting